


# Advanced Certification in AIML
## A Program by IIIT-H and TalentSprint

## Problem Statement

The problem is to identify the author of a  book from a given list of possible authors.

## Learning Objectives

At the end of the experiment, you will be able to :

* Preprocess the text
* Write an algorithm to identify author of a given book
* Use NLTK package

## Background 

Author identification is the task of identifying the author of a given text. It can be considered as a typical classification problem, where a set of books with known authors are used for training. The aim is to automatically determine the corresponding author of an anonymous text. 

## Setup Steps

In [0]:
#@title Please enter your registration id to start: (e.g. P181900101) { run: "auto", display-mode: "form" }
Id = "P181902118" #@param {type:"string"}


In [0]:
#@title Please enter your password (normally your phone number) to continue: { run: "auto", display-mode: "form" }
password = "8860303743" #@param {type:"string"}


In [100]:
#@title Run this cell to complete the setup for this Notebook

from IPython import get_ipython
ipython = get_ipython()
  
notebook="BLR_M1_Mini_Hackathon_AuthorIdentification" #name of the notebook
Answer = "This notebook is graded by mentors on the day of hackathon"
def setup():
#  ipython.magic("sx pip3 install torch")
#    ipython.magic("sx wget https://www.dropbox.com/s/9xivz2pox1i83td/AIML_DS_GOOGLENEWS-VECTORS-NEGATIVE-300_STD.bin?dl=1")
#    ipython.magic("sx mv AIML_DS_GOOGLENEWS-VECTORS-NEGATIVE-300_STD.bin?dl=1 AIML_DS_GOOGLENEWS-VECTORS-NEGATIVE-300_STD.bin")
   print ("Setup completed successfully")
   return

def submit_notebook():
    
    ipython.magic("notebook -e "+ notebook + ".ipynb")
    
    import requests, json, base64, datetime

    url = "https://dashboard.talentsprint.com/xp/app/save_notebook_attempts"
    if not submission_id:
      data = {"id" : getId(), "notebook" : notebook, "mobile" : getPassword()}
      r = requests.post(url, data = data)
      r = json.loads(r.text)

      if r["status"] == "Success":
          return r["record_id"]
      elif "err" in r:        
        print(r["err"])
        return None        
      else:
        print ("Something is wrong, the notebook will not be submitted for grading")
        return None

    elif getAnswer() and getComplexity() and getAdditional() and getConcepts():
      f = open(notebook + ".ipynb", "rb")
      file_hash = base64.b64encode(f.read())

      data = {"complexity" : Complexity, "additional" :Additional, 
              "concepts" : Concepts, "record_id" : submission_id, 
              "answer" : Answer, "id" : Id, "file_hash" : file_hash,
              "notebook" : notebook}

      r = requests.post(url, data = data)
      r = json.loads(r.text)
      print("Your submission is successful.")
      print("Ref Id:", submission_id)
      print("Date of submission: ", r["date"])
      print("Time of submission: ", r["time"])
      print("View your submissions: https://iiith-aiml.talentsprint.com/notebook_submissions")
      print("For any queries/discrepancies, please connect with mentors through the chat icon in LMS dashboard.")
      return submission_id
    else: submission_id
    

def getAdditional():
  try:
    if Additional: return Additional      
    else: raise NameError('')
  except NameError:
    print ("Please answer Additional Question")
    return None

def getComplexity():
  try:
    return Complexity
  except NameError:
    print ("Please answer Complexity Question")
    return None
  
def getConcepts():
  try:
    return Concepts
  except NameError:
    print ("Please answer Concepts Question")
    return None

def getAnswer():
  try:
    return Answer
  except NameError:
    print ("Please answer Question")
    return None

def getId():
  try: 
    return Id if Id else None
  except NameError:
    return None

def getPassword():
  try:
    return password if password else None
  except NameError:
    return None

submission_id = None
### Setup 
if getPassword() and getId():
  submission_id = submit_notebook()
  if submission_id:
    setup()
  
else:
  print ("Please complete Id and Password cells before running setup")



Setup completed successfully


### NOTE: Please feel free to use ML libraries such as Sklearn, NLTK etc whereever applicable

## Stage 1: Sentence Level Stats

You can choose any two authors books from the below list 

* Shakespeare
* Austen
* Chesterton 

The books of the authors are mentioned in **Gutenberg Corpus**.

Click on the [link](https://www.nltk.org/book/ch02.html) for reference.

### Downloading the required nltk Packages

In [101]:
import nltk
nltk.download('gutenberg')
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('stopwords')

[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

### a) Access a particular book by the author chosen by you above. Find number of sentences and words (tokens) present in that book: 1 Mark

Instructions on accessing the Corpus can be found at the following [link](https://www.nltk.org/book/ch02.html)

In [102]:
#####YOUR CODE HERE###############
from nltk.corpus import gutenberg

tokensFromCorpus=gutenberg.words(fileids='shakespeare-hamlet.txt')
print("Number of words in Hamlet - ", len(tokensFromCorpus))

hamlet_sentences = gutenberg.sents(fileids="shakespeare-hamlet.txt")
print("Number of Sentences in Hamlet - ", len(hamlet_sentences))

############################

Number of words in Hamlet -  37360
Number of Sentences in Hamlet -  3106


#Before moving ahead choose your two authors based on the your team-nbr: <br/>
**1.** <br/>
Team=1,6 &nbsp;&nbsp;&nbsp;         Author-A Vs Author-B <br />
Team=2,7 &nbsp;&nbsp;&nbsp;         Author-B Vs Author-C <br />
Team=3,8 &nbsp;&nbsp;&nbsp;         Author-C Vs Author-D <br />
Team=4,5 &nbsp;&nbsp;&nbsp;         Author-D Vs Author-E <br />
 <br />

**2.** Link for each of the author to be used for your problem: <br />

Author-A &nbsp;&nbsp; G.K.Chesterton <br />

*   Author-A -> Rudyard Kipling ->  [Short Stories Collection](http://www.gutenberg.org/files/2781/2781-0.txt) &nbsp;&nbsp;
*   Author-B -> Anton Chekhov [Short Stories Collection](http://www.gutenberg.org/files/1732/1732-0.txt) &nbsp;&nbsp;
*   Author-C -> Guy De Maupassant [Short Stories Collection](http://www.gutenberg.org/cache/epub/21327/pg21327.txt)&nbsp;&nbsp;
*   Author-D -> O Henry [Short Stories Collection](http://www.gutenberg.org/cache/epub/2141/pg2141.txt) &nbsp;&nbsp;
*   Author-E -> Mark Twain [Short Stories Collection](http://www.gutenberg.org/files/245/245-0.txt)&nbsp;&nbsp;


**Downloading raw text from Gutenberg :**
 Hint: Refer section "Electronic Books" in the following  [link](https://www.nltk.org/book/ch03.html) for the instructions.  
 
 
 **Ensure you appropriately split the above text into multiple short stories, which will be your training data - 2 Marks  (Hint: You may use raw.find() and raw.rfind() in the  [link](https://www.nltk.org/book/ch03.html)   to find appropriate index of the start and end location)**

In [103]:
############################
from urllib import request
urlAnton = "http://www.gutenberg.org/files/1732/1732-0.txt"
responseAnton = request.urlopen(urlAnton)
rawAnton = responseAnton.read().decode('utf8').lower()
print(type(rawAnton))
print(len(rawAnton))
print(rawAnton[:75], "\n\n")

############################
urlGuy = "http://www.gutenberg.org/cache/epub/21327/pg21327.txt"
responseGuy = request.urlopen(urlGuy)
rawGuy = responseGuy.read().decode('utf8').lower()
print(type(rawGuy))
print(len(rawGuy))
print(rawGuy[:100], "\n\n")

############################

<class 'str'>
376627
project gutenberg’s the schoolmistress and other stories, by anton chekhov 


<class 'str'>
613359
the project gutenberg ebook of the works of guy de maupassant, vol. 1 (of
8), by guy   de maupassa 




In [0]:
import string
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

def dataCleaningWords(wordList):
    
    tmpList = []
    # 1. Removal of Punctuation Marks 
    tmpList = [each for each in wordList if each not in string.punctuation]
    
    # 2. Removal of Stopwords
    tmpList2 = []
    tmpList2 = [each for each in tmpList if each not in stopwords.words('english')]

    # 3. Lemmatisation 
    tmpList3 = []
    tmpList3 = [WordNetLemmatizer().lemmatize(each, pos="v") for each in tmpList2]
    return tmpList3
  
  
def dataCleaningSent(sentList):
  cleanedSentList = []
  for eachSent in sentList:
    cleanedSentList.append(' '.join(dataCleaningWords(word_tokenize(eachSent))))
  return cleanedSentList

In [118]:
import nltk, re, pprint
from nltk import word_tokenize, sent_tokenize

# from the contents - get the title of all short stories
# "End of the Project" - will make the ending
step = 2

############################

indexOfContentsAnton = rawAnton.find("contents")
indexOfEndAnton = rawAnton.find("End of the Project Gutenberg".lower())

# print(indexOfContents)
# print(indexOfEnd)
shortRawAnton = rawAnton[indexOfContentsAnton:indexOfEndAnton]
wordTokensAnton = word_tokenize(shortRawAnton)
print("len wordTokensAnton", len(wordTokensAnton))

sentTokensAnton = sent_tokenize(shortRawAnton) 
print("len sentTokensAnton", len(sentTokensAnton))

# cleaning
cleanedWordTokensAnton = dataCleaningWords(wordTokensAnton)
cleneadSentTokensAnton = dataCleaningSent(sentTokensAnton)

print("len(cleanedWordTokensAnton) = ", len(cleanedWordTokensAnton))
print("len(cleneadSentTokensAnton) = ", len(cleneadSentTokensAnton))


antonStories = []
# every 150 sentences is a story
for x in range(0, len(cleneadSentTokensAnton), step):
    sentences = cleneadSentTokensAnton[x:x+step]
    antonStories.append(sentences)

print("len(antonStories) = ", len(antonStories), "\n\n")





#########################################



indexOfContentsGuy = rawGuy.find("contents")
indexOfEndGuy = rawGuy.find("FOOTNOTES".lower())

# print(indexOfContentsGuy)
# print(indexOfEndGuy)
shortRawGuy = rawGuy[indexOfContentsGuy:indexOfEndGuy]
wordTokensGuy = word_tokenize(shortRawGuy) 
print("len tokensGuy", len(wordTokensGuy))

sentTokensGuy = sent_tokenize(shortRawGuy) 
print("len sentTokensGuy", len(sentTokensGuy))

# cleaning
cleanedSentTokensGuy = dataCleaningSent(sentTokensGuy)
cleanedWordTokensGuy = dataCleaningWords(wordTokensGuy)

print("len(cleanedWordTokensGuy) = ", len(cleanedWordTokensGuy))
print("len(cleanedSentTokensGuy) = ", len(cleanedSentTokensGuy))

guyStories = []
# every 150 sentences is a story
for x in range(0, len(cleanedSentTokensGuy), step):
    sentences = cleanedSentTokensGuy[x:x+step]
    guyStories.append(sentences)
    
print("len(guyStories) = ", len(guyStories))


############################

len wordTokensAnton 77215
len sentTokensAnton 2970
len(cleanedWordTokensAnton) =  33582
len(cleneadSentTokensAnton) =  2970
len(antonStories) =  1485 


len tokensGuy 126336
len sentTokensGuy 5453
len(cleanedWordTokensGuy) =  52340
len(cleanedSentTokensGuy) =  5453
len(guyStories) =  2727


## Stage 2: Experiment with Handcrafted features representation

**Stylometry: ** is the quantitative study of literary style through computational distant reading methods. It is based on the observation that authors tend to write in relatively consistent, recognizable and unique ways. For example:

Each person has their own unique vocabulary, sometimes rich, sometimes limited. Although a larger vocabulary is usually associated with literary quality, this is not always the case. Ernest Hemingway is famous for using a surprisingly small number of different words in his writing,1 which did not prevent him from winning the Nobel Prize for Literature in 1954.
Some people write in short sentences, while others prefer long blocks of text consisting of many clauses.
No two people use semicolons, em-dashes, and other forms of punctuation in the exact same way.




**You may explore the following ways to analyze the text and generate handcrafted features by searching text in a probing way:**

a) Could  style of punctuation usage help as an handcrafted feature? Both by those who follow punctuations and by those who don't? Interesting [link](https://qwiklit.com/2014/03/05/top-10-authors-who-ignored-the-basic-rules-of-punctuation/) 

b) The same word can sometimes be used in different contexts repeatedly by different authors. Could this fact be converted as a handcrafted feature? [link](https://www.nltk.org/book/ch01.html)

c) The above two are merely examples; As you might have noticed already the NLTK book [link](https://www.nltk.org/book/) offers several methods of analyzing and understanding text. Each of these analyses is in itself capable of being an handcrafted feature. **However for your evaluation a minimal set of useful handcrafted features which is helping you prove a classification of an is sufficient**

d) Could most command words be used to distinguish authors?  Refer "Counting Vocabulary" section of the [link](https://www.nltk.org/book/ch01.html)

e) How about using a count of most frequently used bi-gram, tri-grams and using it to classify an author?

f) How about using the frequency historgram of the most frequently used words across the stories by a given author a useful feature? 

The limit here is endless limited only by your imagination, and of course your accuracy ! :)


### a) Identify 10 handcrafted features of the kind described by the  story above- 2 Marks
 




In [0]:
# UniqueWords    AvgSentLenth  mcwFeature   Label 

from nltk.probability import FreqDist
finalVectorCombined = []

def buildFinalVectorCombined(storyList, label):
  for eachStory in storyList:
    sentLenSum = 0
    wordsList = []
    for eachSent in eachStory:
      wordsList.extend(word_tokenize(eachSent))
      sentLenSum += len(word_tokenize(eachSent))

    avgSentLength = sentLenSum/len(eachStory)
    uniqueWordsLength = len(set(wordsList))

    fdist = FreqDist(wordsList)
    mcw = sum([y for x,y in fdist.most_common(50)])
    mcwFeature = 0
    if(len(wordsList)!=0):
      mcwFeature = mcw/len(wordsList)

    # print("avgSentLength =", avgSentLength, "uniqueWordsLength = ", uniqueWordsLength, "mcwFeature = ", mcwFeature)
    finalVectorCombined.append([uniqueWordsLength,avgSentLength, mcwFeature, label])
    
buildFinalVectorCombined(antonStories, 0)
buildFinalVectorCombined(guyStories, 1)

In [115]:
print("len(finalVectorCombined) = ", len(finalVectorCombined))
print(finalVectorCombined[:5])
print(finalVectorCombined[-5:])


len(finalVectorCombined) =  4212
[[47, 25.0, 1.0, 0], [40, 20.5, 1.0, 0], [39, 22.5, 1.0, 0], [34, 18.5, 1.0, 0], [30, 17.5, 1.0, 0]]
[[6, 3.0, 1.0, 1], [4, 2.0, 1.0, 1], [4, 2.0, 1.0, 1], [19, 10.0, 1.0, 1], [10, 10.0, 1.0, 1]]


### b) Write functions for generated 8 of the above 10 features you described above  - 4 Marks

## Stage 3: Create a  sentence level text processing and representation (to be used in the next stage for classification):
TFIDF or CountVectorizer or Word2vec features 


### a) Performing basic cleanup operations such as removing the newline characters and removing trailing spaces - 2 Marks
For example: If you text looks as follows \[' This is a sentence\n\r. Another sentence \n'], after newline removal your sentence would look like \['This is sentence a . Another sentence'] . In order to do this you can try using a combination of split() and join()

In [0]:
##################### YOUR CODE HERE ##################################

### b) Tokenize the sentences -  2 Marks
**Example of a tokenized sentence:** <br/>Using same example from the previous cell, after tokenization the sentence <br/> \['This is sentence a . Another sentence']  would now be split into two sentences i.e. two elements in the list as follows:  \['This is a sentence .' , ' Another sentence']  
<br/>
You may further choose to break sentences to individual words **if needed. **<br/>
**Example of sentences as list of words:**<br/>
**Before:** ['This is a sentence .' , ' Another sentence']<br/>
**After:** ['This', 'is' ,'a', 'sentence' , ' . ' , ' Another ', ' sentence ' ]

Hint: Consider using 'List comprehensions', whereby you split the sentence with spaces (using split('')) and populate the resulting output a list comprehension.



### c) Is stop word removal necessary in the context of author identification? Your thoughts below? 1 Mark

In [0]:
###### YOUR ANSWER IN TEXT##########



### c) Construct a labelled dataset of sentences - 3 Marks

To create a representation of text, convert it into vectors (numbers).** Use any one ** of the following algorithms for the task :

* Countvectorizer or
* TFIDFVectorizer or 
* Word2Vec 

References Documents: 

1.   [Countvectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html)
2.  [TFIDFVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html)


In [119]:
############################ YOUR CODE HERE###########################
#Define a vectorizer
#fit_transform your data to teh vectorizer
#Convert to numpy array if needed


# Building STEP 3
# TfidfVectorizer


# TfidfVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
mergedlist = cleneadSentTokensAnton + cleanedSentTokensGuy
print(len(mergedlist))
vectorizer = TfidfVectorizer(stop_words = 'english', ngram_range = (1,3))
Step3Vector = vectorizer.fit_transform(mergedlist)
print(vectorizer.get_feature_names())
print(Step3Vector.shape)

8423
['10', '10 188', '10 188 non', '10 happen', '10 happen morrow', '10 rouen', '10 rouen hotel', '10 surprise', '10 surprise ordeals', '10 weave', '10 weave table', '11', '11 dint', '11 dint despise', '11 stop', '11 stop home', '11 time', '11 time paul', '117', '117 umbrella', '117 umbrella 131', '12', '12 chimneysweep', '12 chimneysweep black', '12 clock', '12 clock night', '12 paris', '12 slight', '12 slight feverish', '13', '13 attack', '13 attack certain', '13 make', '13 make slightest', '131', '131 uncle', '131 uncle sosthenes', '14', '14 1870', '14 1870 end', '14 1885', '14 1885 slightest', '14 _fête_', '14 _fête_ republic', '14 lose', '14 moreau', '14 moreau wife', '143', '15', '15 certainly', '15 certainly way', '15 mont', '15 mont saint', '152', '152 philosopher', '152 philosopher 162', '16', '16 ill', '16 ill decidedly', '16 make', '16 make tell', '16 manage', '16 manage escape', '16 saw', '16 saw things', '162', '162 lock', '162 lock door', '17', '17 oh', '17 strange', '17

In [131]:
# Building STEP 3
# CountVectorizer
from sklearn.feature_extraction.text import CountVectorizer
mergedlist = cleneadSentTokensAnton + cleanedSentTokensGuy

vectorizer = CountVectorizer()
Step3Vector = vectorizer.fit_transform(mergedlist)
print(vectorizer.get_feature_names())
print(Step3Vector.shape)

['10', '11', '117', '12', '13', '131', '14', '143', '15', '152', '16', '162', '17', '171', '179', '18', '1870', '1871', '1874', '188', '1882', '1885', '19', '190', '198', '20', '206', '21', '213', '221', '228', '25', '263', '270', '279', '286', '293', '30', '303', '311', '321', '336', '343', '350', '358', '376', '54', '76th', '82', '8th', '_a', '_ad', '_at', '_august', '_benedictus_', '_bock_', '_brisque_', '_casque', '_charity_', '_che', '_châteaux_', '_city', '_cranium_', '_crash', '_crevasses_', '_curé_', '_delila_', '_etienne_', '_exactly_', '_fame_', '_folâtre_', '_fête_', '_gil', '_good', '_géant', '_handsome', '_hide', '_him_', '_his', '_humanity_', '_i_', '_impossible', '_in', '_ingenue_', '_journal', '_july', '_june', '_la', '_lycée_', '_mademoiselle', '_mademoiselle_', '_madness_', '_man_', '_manifestation_', '_manifesting_', '_market', '_maternelle_', '_matilda_', '_may', '_messenger', '_mica', '_mica_', '_monde_', '_monsieur', '_no_', '_non', '_patito_', '_puss', '_realizat

## Stage 4: Classification : Marks - 3
Perform a classification using either features obtained from Step2 or Step3

In [0]:
##############################################YOUR CODE HERE###############################
# Using STEP 3

from sklearn.model_selection import train_test_split

y = [0 for each in sentTokensAnton]
y.extend([1 for each in sentTokensGuy])
X_train, X_test, y_train, y_test = train_test_split(Step3Vector, y, test_size=0.2, random_state=1234)



In [0]:

##############################################YOUR CODE HERE###############################
# Using STEP 2

import pandas as pd
from sklearn.model_selection import train_test_split

df = pd.DataFrame(finalVectorCombined)
df.columns = ['A', 'B', 'C', 'D']
X = df[['A', 'B', 'C']]
y = df[['D']]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1234)



In [133]:
#################
# Decision Tree #
#################

from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(max_depth=1)
decision_tree.fit(X_train,y_train)
decision_tree.predict(X_test)
print(decision_tree.score(X_test,y_test))



#############
###  KNN  ###
#############

from sklearn.neighbors import KNeighborsClassifier

neigh = KNeighborsClassifier(n_neighbors=3)
# Training or fitting the model with the train data
neigh.fit(X_train,y_train)
# Testing the trained model
neigh.predict(X_test)
# Calculating the score
print(neigh.score(X_test,y_test))



##########################
###  Linear Classifier  ##
##########################

from sklearn import linear_model

linear_classifier = linear_model.SGDClassifier()
linear_classifier.fit(X_train,y_train)
linear_classifier.predict(X_test)
print(linear_classifier.score(X_test,y_test))

0.6623145400593472
0.6545994065281899
0.7833827893175074


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/stochastic_gradient.py:166: FutureWarning: max_iter and tol parameters have been added in SGDClassifier in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  FutureWarning)


**Using step 2 vector:**


---
0.6500593119810202

0.608540925266904

0.6524317912218268

---



**Using step 3 vector: **

------Using TFIDFVectorizer

0.6623145400593472

0.6534124629080119

0.799406528189911

---

------Using CountVectorizer*

0.6623145400593472

0.6545994065281899

0.7833827893175074

---


# Further Ideas for exploration after the hackathon:

**Statistical analysis** of text using NLP, by analysis meaning of sentences, feature based grammars and analyzing structure of sentences! For all of this you can reference www.nltk.org/book

### Please answer the questions below to complete the experiment:

In [0]:
#@title How was the experiment? { run: "auto", form-width: "500px", display-mode: "form" }
Complexity = "Good and Challenging for me" #@param ["Too Simple, I am wasting time", "Good, But Not Challenging for me", "Good and Challenging for me", "Was Tough, but I did it", "Too Difficult for me"]


In [0]:
#@title If it was very easy, what more you would have liked to have been added? If it was very difficult, what would you have liked to have been removed? { run: "auto", display-mode: "form" }
Additional = " good" #@param {type:"string"}

In [0]:
#@title Can you identify the concepts from the lecture which this experiment covered? { run: "auto", vertical-output: true, display-mode: "form" }
Concepts = "Yes" #@param ["Yes", "No"]

In [127]:
#@title Run this cell to submit your notebook for grading { vertical-output: true }
try:
  if submission_id:
      return_id = submit_notebook()
      if return_id : submission_id =return_id
  else:
      print("Please complete the setup first.")
except NameError:
  print ("Please complete the setup first.")

Your submission is successful.
Ref Id: 3211
Date of submission:  30 Mar 2019
Time of submission:  16:04:18
View your submissions: https://iiith-aiml.talentsprint.com/notebook_submissions
For any queries/discrepancies, please connect with mentors through the chat icon in LMS dashboard.
